Implement server-side rendering and Redis caching layer

dweinzimmer opened the pull request:
Tickets: https://trello.com/c/vDJhwKMz/38-proxy and https://trello.com/c/qiedQcZL/34-add-a-caching-layer-redis

This pull request implements the completed proxy server with server-side rendering and a Redis caching layer. The proxy server requests server-side-rendered objects of `{ssr_html, props}` from each service and injects that into a standard template, which includes all necessary non-dynamic links (CSS files, React and ReactDOM, client bundles, and `ReactDOM.hydrate` statements for each microservice). 

In addition, past results are cached in a Redis database that resides on a separate EC2 instance. The `cache` function defined on line 36 of `server.js` is invoked as middleware on line 192 of the same file, serving cached results if they are available, and continuing on if not.

Let me know if you have any thoughts!

dweinzimmer committed:
Modify proxy server for server-side rendering

The server does not yet run because it is dependent upon
reconfigurations to the client. Specifically, the client will be
modified to create the destination:Node bundle required by this
proxy server.

dweinzimmer committed:
Save progress on SSR

dweinzimmer committed:
Complete SSR on Proxy, incorporating only Reviews service so far

dweinzimmer committed:
Modify Proxy to display SSR, currently with Reviews only

dweinzimmer committed:
Create Dockerfile

dweinzimmer committed:
Update proxy server with other microservices

dweinzimmer committed:
Update to include all files for Description and Bookings services

dweinzimmer committed:
Update to include Neighborhood microservice

dweinzimmer committed:
Implement Redis caching

dweinzimmer committed:
Add Lighthouse command line testing

kchen395 reviewed (approved):
In your getSSRObjects function, you could create a helper function for your axios get requests to reduce repeated code. Besides that, good work!

dweinzimmer committed:
Factor out 'getServices' helper function and accept empty inputs

Per Kai's code review (PR#1), factor out the repetitive code in
requests to all four services to a getServices function. Also,
refactored to accept empty inputs from services that are not online.

dweinzimmer commented:
Hi @kchen395, thank you! I factored out that repetitive code into a `getServices` helper function, and also refactored the injection of SSR html & props to allow for undefined inputs to not cause the client to hang. 

Thanks for the code review!
